consensus/propagation/catchup.go

```diff
@@ -1,54 +1,108 @@
-		if !found {
-			blockProp.Logger.Error("couldn't find proposal", "height", height, "round", -1)
+		if height == currentHeight && round == currentRound {
 			continue
 		}
-		if partSet.BitArray() == nil {
-			panic("partSet is nil when getting proposal!")
-		}
-		missingParts := partSet.BitArray().Not()
-		wantPart := &proptypes.WantParts{
-			Parts:  missingParts,
-			Height: height,
-			Round:  heightProposal.Round,
-		}
-		blockProp.broadcastWants(wantPart, blockProp.self)
-	}
 
-	// if we're missing rounds for the current height, request them from all the peers
-	for round := int32(0); round < targetRound; round++ {
-		_, partSet, found := blockProp.GetProposal(targetHeight, round)
-		if !found {
-			blockProp.Logger.Error("couldn't find proposal", "height", targetHeight, "round", round)
+		if prop.block.IsComplete() {
 			continue
 		}
-		if partSet.BitArray() == nil {
-			panic("partSet is nil when getting proposal!")
+
+		// only re-request original parts that are missing, not parity parts.
+		missing := prop.block.BitArray().Not()
+		if missing.IsEmpty() {
+			// this should never be hit due to the check above.
+			continue
 		}
-		missingParts := partSet.BitArray().Not()
-		wantPart := &proptypes.WantParts{
-			Parts:  missingParts,
-			Height: targetHeight,
-			Round:  round,
+
+		// make requests from different peers
+		peers = shuffle(peers)
+
+		for _, peer := range peers {
+			mc := missing.Copy()
+			reqs, has := peer.GetRequests(height, round)
+			if has {
+				mc = mc.Sub(reqs)
+			}
+
+			if mc.IsEmpty() {
+				continue
+			}
+
+			e := p2p.Envelope{
+				ChannelID: WantChannel,
+				Message: &protoprop.WantParts{
+					Parts:  *missing.ToProto(),
+					Height: height,
+					Round:  round,
+					Prove:  true,
+				},
+			}
+
+			if !p2p.TrySendEnvelopeShim(peer.peer, e, blockProp.Logger) { //nolint:staticcheck
+				blockProp.Logger.Error("failed to send want part", "peer", peer, "height", height, "round", round)
+				continue
+			}
+
+			// keep track of which requests we've made this attempt.
+			missing.Sub(mc)
+			peer.AddRequests(height, round, missing)
 		}
-		blockProp.broadcastWants(wantPart, blockProp.self)
 	}
 }
+
+func (blockProp *Reactor) AddCommitment(height int64, round int32, psh *types.PartSetHeader) {
+	blockProp.pmtx.Lock()
+	defer blockProp.pmtx.Unlock()
+
+	if blockProp.proposals[height] == nil {
+		blockProp.proposals[height] = make(map[int32]*proposalData)
+	}
+
+	combinedSet := proptypes.NewCombinedPartSetFromOriginal(types.NewPartSetFromHeader(*psh), true)
+
+	if blockProp.proposals[height][round] != nil {
+		return
+	}
+
+	blockProp.proposals[height][round] = &proposalData{
+		compactBlock: &proptypes.CompactBlock{
+			Proposal: types.Proposal{
+				Height: height,
+				Round:  round,
+			},
+		},
+		catchup:     true,
+		block:       combinedSet,
+		maxRequests: bits.NewBitArray(int(psh.Total * 2)), // this assumes that the parity parts are the same size
+	}
+}
```

Comment:
this allows the consensus reactor to add a partset header each time the node falls behind and then sees a valid commit later